cpp/src/slice2py/PythonUtil.cpp

```diff
@@ -2563,223 +2539,99 @@ Slice::Python::printHeader(IceInternal::Output& out)
-                    if (subArg == "tuple" || subArg == "list" || subArg == "default")
-                    {
-                        continue;
-                    }
-                    else if (subArg == "array.array" || subArg == "numpy.ndarray" || subArg.find("memoryview") == 0)
-                    {
-                        // The memoryview sequence metadata is only valid for integral builtin
-                        // types excluding strings.
-                        BuiltinPtr builtin = dynamic_pointer_cast<Builtin>(seq->type());
-                        if (builtin)
-                        {
-                            switch (builtin->kind())
-                            {
-                                case Builtin::KindBool:
-                                case Builtin::KindByte:
-                                case Builtin::KindShort:
-                                case Builtin::KindInt:
-                                case Builtin::KindLong:
-                                case Builtin::KindFloat:
-                                case Builtin::KindDouble:
-                                {
-                                    continue;
-                                }
-                                default:
-                                {
-                                    break;
-                                }
-                            }
-                        }
-                    }
-                }
+                return "the 'python:package' metadata can only be applied at the file level or to top-level modules";
             }
-            ostringstream msg;
-            msg << "ignoring invalid metadata '" << *s << "'";
-            type->unit()->warning(s->file(), s->line(), InvalidMetadata, msg.str());
-            newMetadata.remove(s);
-        }
-    }
-    return newMetadata;
-}
+            return nullopt;
+        },
+    };
+    knownMetadata.emplace("python:package", std::move(packageInfo));
 
-void
-Slice::Python::MetadataVisitor::reject(const ContainedPtr& cont)
-{
-    MetadataList localMetadata = cont->getMetadata();
+    // "python:pkgdir"
+    MetadataInfo pkgdirInfo = {
+        .validOn = {typeid(Unit)},
+        .acceptedArgumentKind = MetadataArgumentKind::RequiredTextArgument,
+    };
+    knownMetadata.emplace("python:pkgdir", std::move(pkgdirInfo));
+
+    // "python:seq"
+    // We support 3 arguments to this metadata: "default", "list", and "tuple".
+    // We also allow users to omit the "seq" in the middle, ie. "python:seq:list" and "python:list" are equivalent.
+    MetadataInfo seqInfo = {
+        .validOn = {typeid(Sequence)},
+        .acceptedArgumentKind = MetadataArgumentKind::SingleArgument,
+        .validArgumentValues = {{"default", "list", "tuple"}},
+        .acceptedContext = MetadataApplicationContext::DefinitionsAndTypeReferences,
+    };
+    MetadataInfo unqualifiedSeqInfo = {
```

Comment:
It would be clearer to define this variable after:
```cpp
knownMetadata.emplace("python:seq", std::move(seqInfo));
```

Reply:
Fine, I moved it.